Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+
+```proto
+enum LogLevel {
+  trace = 0;
+  // omitted
+  fatal = 5;
+}
+message LogMessage {
+  string message = 1; // A message for the log
+  LogLevel level = 2; // Log level
+  // omitted
+}
+
+service Lumberjack {
+  // Take a LogMessage object and an empty message
+  rpc SendLog (LogMessage) returns (google.protobuf.Empty);
+}
+```
+
+##### Client
+Tazama abstracts away the need to write your own [gRPC] client. Simply initialising a logger and providing it with an address will create a [gRPC] client behind the scenes which will send requests to the address that was provided.
+
+An example:
+
+```js
+// No sidecar host is provided in the LoggerService constructor
+const logger = new LoggerService();
+```
+
+An alternative implementation, **opting into** logging with the sidecar (as well as [lumberjack]) may be:
+```js
+const logger = new LoggerService("192.0.0.1:1234");
+```
+> [!NOTE]  
+> If a sidecar host is provided in the `LoggerService` constructor, you have to ensure your sidecar is live on the address you provided
+
+As [gRPC] is in use, it allows Tazama to be flexible enough to allow logs coming from any client (in any programming language) as long as it adheres to the type that the server expects (defined in the protobuf format above).
+
+After receiving the logs, the [event-sidecar] sends them to [lumberjack] via [NATS]
+
+#### Lumberjack
+Lumberjack is a microservice that receives all log events before redirecting them to a central place. This microservice utilises [NATS] to listen for incoming messages with a specific subject. The [event-sidecar] sends messages with a specific subject that [lumberjack] listens for. This microservice depends on [pino-elasticsearch](https://github.com/pinojs/pino-elasticsearch), which is what acts as our transport to transform our logs from our wire format (in the protofile), to the format that [Elastic] search expects.
+
+This microservice has an additional responsibility of batching the received logs before transmission. This has a wide range of benefits as ultimately, I/O operations are reduced which will generally have implications on from improved performance coming from less network calls which can also affect costs in cloud environments.
+
+### Implementation
+The [event-sidecar] and [lumberjack] are optional components in implementation. Perhaps hardware is limited or there may be some situations whereby running more microservices is not ideal, Tazama will still have a functional logging implementation if the [event-sidecar] and [lumberjack] are excluded from deployment. As mentioned before, you can opt in to using the [event-sidecar] and [lumberjack] for logging by specifying the sidecar host when initialising the logger service.
+
+#### Configuration
+
+##### With Event-Sidecar and Lumberjack
```

Comment:
Please provide a summary of what this means.

Might be worth making another note that both of these are required. Also, following the "sequence of operation", I'd document the side-car before lumberjack.